internal/controller/controller_functions.go

```diff
@@ -145,6 +133,260 @@ func (r *BaseReconciler) CreateEventFilter() predicate.Funcs {
+	var maxDelay *int32
+	var maxDelayRule *karov1alpha1.RestartRule
+
+	for _, rule := range rules {
+		if rule.Spec.DelayRestart != nil {
+			if maxDelay == nil || *rule.Spec.DelayRestart > *maxDelay {
+				maxDelay = rule.Spec.DelayRestart
+				maxDelayRule = rule
+			}
+		}
+	}
+
+	return maxDelay, maxDelayRule
+}
+
+// getTargetSpec extracts the target spec for a given target key
+func (r *BaseReconciler) getTargetSpec(tgtKey targetKey, rules []*karov1alpha1.RestartRule) karov1alpha1.TargetSpec {
+	firstRule := rules[0]
+	var targetSpec karov1alpha1.TargetSpec
+
+	for _, target := range firstRule.Spec.Targets {
+		targetNS := target.Namespace
+		if targetNS == "" {
+			targetNS = firstRule.Namespace
+		}
+		if target.Kind == tgtKey.kind && target.Name == tgtKey.name && targetNS == tgtKey.namespace {
+			targetSpec = target
+
+			break
+		}
+	}
+
+	return targetSpec
+}
+
+// extractRuleNames extracts rule names for logging
+func (r *BaseReconciler) extractRuleNames(rules []*karov1alpha1.RestartRule) []string {
+	ruleNames := make([]string, len(rules))
+	for i, rule := range rules {
+		ruleNames[i] = rule.Name
+	}
+
+	return ruleNames
+}
+
+// logProcessingTarget logs information about the target being processed
+func (r *BaseReconciler) logProcessingTarget(logger logr.Logger, tgtKey targetKey, resourceName, resourceType string, ruleNames []string) {
+	logger.Info("Processing restart for target",
+		"target", fmt.Sprintf("%s/%s/%s", tgtKey.kind, tgtKey.namespace, tgtKey.name),
+		"resource", resourceName,
+		"resourceType", resourceType,
+		"restartRules", ruleNames,
+		"ruleCount", len(ruleNames))
+}
+
+// processDelayedRestart handles delayed restart logic
+func (r *BaseReconciler) processDelayedRestart(ctx context.Context, logger logr.Logger, tgtKey targetKey, targetSpec karov1alpha1.TargetSpec, rules []*karov1alpha1.RestartRule, ruleNames []string, maxDelay *int32, maxDelayRule *karov1alpha1.RestartRule, resourceName, resourceType string) {
+	delay := time.Duration(*maxDelay) * time.Second
+
+	targetKey := TargetKey{
+		Kind:      tgtKey.kind,
+		Name:      tgtKey.name,
+		Namespace: tgtKey.namespace,
+	}
+
+	restartFunc := r.createRestartFunc(ctx, logger, targetSpec, rules, ruleNames, targetKey, resourceName, resourceType)
+
+	finalDelay, isNew, err := r.DelayedRestartManager.ScheduleRestart(ctx, targetKey, delay, restartFunc)
+	if err != nil {
+		r.handleScheduleError(ctx, logger, err, targetKey, delay, targetSpec, rules, resourceName, resourceType)
+
+		return
+	}
+
+	r.logScheduleResult(logger, isNew, targetKey, finalDelay, delay, resourceName, resourceType, ruleNames, maxDelayRule)
+}
+
+// createRestartFunc creates the restart function for delayed execution
+func (r *BaseReconciler) createRestartFunc(_ context.Context, logger logr.Logger, targetSpec karov1alpha1.TargetSpec, rules []*karov1alpha1.RestartRule, ruleNames []string, targetKey TargetKey, resourceName, resourceType string) RestartFunc {
```

Comment:
That function definition seems awfully long. TODO: Examine whether all these parameters are really used.